src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,143 @@
+            if (!(scm instanceof BitbucketSCM)) {
+                return null;
+            }
+
+            BitbucketSCM bitbucketSCM = (BitbucketSCM) scm;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCM.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                return null;
+            }
+
+            BitbucketSCMRepository repository = bitbucketSCM.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            return new BitbucketSCMFileSystem(filePathClient, null, bitbucketSCM.getBranches().get(0).toString());
+        }
+
+        @Override
+        public SCMFileSystem build(SCMSource source, SCMHead head,
+                                   @CheckForNull SCMRevision scmRevision) throws IOException, InterruptedException {
+            if (!(source instanceof BitbucketSCMSource)) {
+                return null;
+            }
+            BitbucketSCMSource bitbucketSCMSource = (BitbucketSCMSource) source;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCMSource.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                return null;
+            }
+            BitbucketSCMRepository repository = bitbucketSCMSource.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            if (scmRevision != null && scmRevision.getHead() instanceof GitBranchSCMHead) {
+                return new BitbucketSCMFileSystem(filePathClient, scmRevision, ((GitBranchSCMHead) scmRevision.getHead()).getRef());
+            }
+            // Unsupported ref type. Lightweight checkout not supported
+            return null;
+        }
+
+        @Override
+        public boolean supports(SCM scm) {
+            if (scm instanceof BitbucketSCM) {
```

Comment:
This is really specific and not signified at all by the branch selector. As I have learned, this is in keeping with prior art due to the limitations of the pipeline implementation, but I think this needs to be documented somewhere. Perhaps a readme change?

Reply:
Yeah, and maybe a comment on the feature request? Depending on what you've got in mind in terms of documentation.